nengo_spa/algebras/base.py

```diff
@@ -220,6 +301,66 @@ def implement_binding(self, n_neurons_per_d, d, unbind_left, unbind_right):
         """
         raise NotImplementedError()
 
+    def sign(self, v):
+        """Returns the sign of *v* defined by the algebra.
+
+        The exact definition of the sign depends on the concrete algebra, but
+        should be analogous to the sign of a number in so far that
+
+        * binding of two "positive" vectors produces another "positive" vector,
+        * binding of a "negative" vector with itself produces a "positive"
+          vector,
+        * binding of a "negative" and "positive" vectors produces a "negative"
+          vector.
+
+        Furthermore, if the algebra supports fractional binding powers, it
+        should do so for all "non-negative" vectors, but not "negative" vectors.
+
+        If an algebra does not have the notion of a sign, it may raise a
+        *NotImplementedError*.
+
+        Parameters
+        ----------
+        v : (d,) ndarray
+            Vector to determine sign of.
+
+        Returns
+        -------
+        AbstractSign
+            The sign of the input vector.
+
+        See Also
+        --------
+        AbstractAlgebra.abs
+        """
+        raise NotImplementedError()
+
+    def abs(self, v):
+        """Returns the absolute vector of *v* defined by the algebra.
+
+        The exact definition of "absolute vector" may depend on the concrete
+        algebra. It should be a "positive" vector (see `.sign`) that relates
+        to the input vector.
+
+        The default implementation requires that the possible signs of the
+        algebra correspond to actual vectors within the algebra. It will bind
+        the sign vector (from the left side) to the vector *v*.
+
+        If an algebra does not have the notion of a sign or absolute vector,
+        it may raise a *NotImplementedError*.
+
+        Parameters
+        ----------
+        v : (d,) ndarray
+            Vector to obtain the absolute vector of.
+
+        Returns
+        -------
+        (d,) ndarray
+            The absolute vector relating to the input vector.
+        """
+        return self.bind(self.sign(v).to_vector(len(v)), v)
```

Comment:
Without thinking about it very deeply, it seems more natural to me for the default implementation for `abs` to be something like:
```
if self.sign(v).is_negative():
    v = self.bind(self.negative_identity_element(len(v)), v)
```
i.e., "'multiply' it by -1 if it is negative".

Is that consistent with this definition?

Reply:
Similarly, would there be any use in having a `negate` method that flips the sign of `v`?

Reply:
Oh, no this won't work because there is only one negative identity for HRR but there are 3 cases that need to be handled to make it positive based on which of the two coefficients are -1 (or both).